### PR TITLE
Only load plugins from correct folders

### LINKED ISSUE #15035: Change plugin loading logic
We need to change the plugin loading logic as follows:

 * TS Server should only load plugins from its peer/parent node_modules folders ("local" folders)
 * Hosts can specify additional plugin probing folders ("global" folders)
 * Hosts can also specify global plugins to load; these plugins do not get configuration settings

Open question: Should globally-specified plugin names probe from local folders before global folders? Or probe global folders only?

/cc @mhegazy @mjbvz

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes #15035 

Includes refactoring of constructor signatures because they had way too many arguments.

### CODE REVIEW

File: src/server/project.ts [hunk 1]
```diff
@@ -878,13 +875,34 @@ namespace ts.server {
...
+            // Enable tsconfig-specified plugins
             for (const pluginConfigEntry of options.plugins) {
-                const searchPath = getDirectoryPath(this.configFileName);
-                const resolvedModule = <PluginModuleFactory>Project.resolveModule(pluginConfigEntry.name, searchPath, host, log);
+                this.enablePlugin(pluginConfigEntry, searchPaths);
+            }
+            // Enable global plugins with synthetic configuration entries
+            for (const globalPluginName of this.projectService.globalPlugins) {
+                // Provide global: true so plugins can detect why they can't find their config
+                this.enablePlugin({ name: globalPluginName, global: true } as PluginImport, searchPaths);
```
Comment (billti) [inline src/server/project.ts]: Should we skip loading a global if a local copy of the same plugin is given (as that can have config also).

File: src/server/project.ts [hunk 2]
```diff
@@ -878,13 +875,34 @@ namespace ts.server {
                 return;
             }
 
+            // Search our peer node_modules, then any globally-specified probe paths
+            // ../../.. to walk from X/node_modules/typescript/lib/tsserver.js to X/node_modules/
+            const searchPaths = [combinePaths(host.getExecutingFilePath(), "../../.."), ...this.projectService.pluginProbeLocations];
```
Comment (mhegazy) [inline src/server/project.ts]: i would do all the node-process-specific work in server.ts. we publish this as a library s well, and users like WebStorm might have a different layout.

File: src/server/server.ts [hunk 3]
```diff
@@ -374,34 +388,25 @@ namespace ts.server {
...
-                    typingsInstaller || nullTypingsInstaller,
-                    Buffer.byteLength,
-                    process.hrtime,
-                    logger,
-                    canUseEvents);
-
-                if (telemetryEnabled && typingsInstaller) {
-                    typingsInstaller.setTelemetrySender(this);
-                }
+        constructor(options: IOSessionOptions) {
```
Comment (mhegazy) [inline src/server/server.ts]: you could write this as `constructor({ host, installerEventPort, globalTypingsCacheLocation, typingSafeListLocation, canUseEvents } : IOSessionOptions)`

File: src/server/project.ts (hunk 1 above)
Comment (mhegazy) [inline src/server/project.ts]: i thought we are not doing the globalPlugins at this time?

**Comment by mhegazy** [comment]
// CC: @chuckjaz, this will change how loading for plugins happen. the expectation now is that the LS plugin has to be installed as a VSCode extension **instead of** locally in the project, in addition to specifying the name in the `plugins` list in the `tsconfig.json`

**Comment by chuckjaz** [comment]
@mhegazy It is unclear from this change how `@angular/language-service` gets to the right place. What must the VSCode extension look like to get this to work? Also, how does this effect, say, the sublime plugin. Does it also need special treatment?

**Comment by mhegazy** [comment]
> @mhegazy It is unclear from this change how @angular/language-service gets to the right place. What must the VSCode extension look like to get this to work?

@mjbvz is adding support for a VSCode extension package property to indicate that this is a tsserver plugin. he can provide more details for this scenario.

> Also, how does this effect, say, the sublime plugin. Does it also need special treatment?

We will have to add a similar mechanism for sublime plugin to lookup an setting that would be set by another sublime extension.

**Comment by chuckjaz** [comment]
@mhegazy @mjbvz What is the time-frame for this, 2.3?

If so, I can roll this into my current plugin. It also seems like the plugin should, however, look for the version of `@angular/language-service` in the project. Otherwise, it would be difficult for users to have the version of the language service match their version of the framework without a intricate plugin dance.

**Comment by mhegazy** [comment]
> @mhegazy @mjbvz What is the time-frame for this, 2.3?

yes.

> If so, I can roll this into my current plugin. It also seems like the plugin should, however, look for the version of @angular/language-service in the project. Otherwise, it would be difficult for users to have the version of the language service match their version of the framework without a intricate plugin dance.

VSCode comes in with a backed in TS version. in this case, users will have to install a matching version of the plugin manually.

If users use a custom version of the server, the server will look first in the node_modules folder that contains it, so a custom plugin would be loaded as well.

**Comment by chuckjaz** [comment]
@mhegazy With a custom build of tsserver, will a plugin provided extension still work?

**Comment by mhegazy** [comment]
> @mhegazy With a custom build of tsserver, will a plugin provided extension still work?

yes. the server injects its containing node_module at the top of the lookup locations that the IDE provides. if one is found in that folder it is given precedence, otherwise, installed plugin is loaded.

**Pull request merged**